src/OpenSage.Game/Scene3D.cs

```diff
@@ -84,45 +90,204 @@ internal IEnumerable<AttachedParticleSystem> GetAllAttachedParticleSystems()
+            HeightMap heightMap,
+            MapObject[] mapObjects,
+            List<Team> teams,
+            out WaypointCollection waypointCollection,
+            out GameObjectCollection gameObjects,
+            out RoadCollection roads,
+            out Bridge[] bridges)
+        {
+            var waypoints = new List<Waypoint>();
+            gameObjects = AddDisposable(new GameObjectCollection(contentManager));
+            var roadsList = new List<Road>();
+            var bridgesList = new List<Bridge>();
+
+            var roadTopology = new RoadTopology();
+
+            for (var i = 0; i < mapObjects.Length; i++)
+            {
+                var mapObject = mapObjects[i];
+
+                var position = mapObject.Position;
+
+                switch (mapObject.RoadType & RoadType.PrimaryType)
+                {
+                    case RoadType.None:
+                        switch (mapObject.TypeName)
+                        {
+                            case "*Waypoints/Waypoint":
+                                waypoints.Add(new Waypoint(mapObject));
+                                break;
+
+                            default:
+                                position.Z += heightMap.GetHeight(position.X, position.Y);
+
+                                var gameObject = GameObject.FromMapObject(mapObject, teams, contentManager, gameObjects, position);
+                                if (gameObject != null)
+                                {
+                                    gameObjects.Add(gameObject);
+                                }
+
+                                break;
+                        }
+                        break;
+
+                    case RoadType.BridgeStart:
+                    case RoadType.BridgeEnd:
+                        // Multiple invalid bridges can be found in e.g GLA01.
+                        if ((i + 1) >= mapObjects.Length || !mapObjects[i + 1].RoadType.HasFlag(RoadType.BridgeEnd))
+                        {
+                            Logger.Warn($"Invalid bridge: {mapObject.ToString()}, skipping...");
+                            continue;
+                        }
+
+                        var bridgeEnd = mapObjects[++i];
+
+                        bridgesList.Add(AddDisposable(new Bridge(
+                            contentManager,
+                            heightMap,
+                            mapObject,
+                            mapObject.Position,
+                            bridgeEnd.Position,
+                            gameObjects)));
+
+                        break;
+
+                    case RoadType.Start:
+                    case RoadType.End:
+                        var roadEnd = mapObjects[++i];
+
+                        // Some maps have roads with invalid start- or endpoints.
+                        // We'll skip processing them altogether.
+                        if (mapObject.TypeName == "" || roadEnd.TypeName == "")
```

Comment:
Maybe `String.IsNullOrEmpty` might be a bit more readable? Or checking for `Length == 0`